### ISSUE TITLE
Request: Functionality for obtaining gradients that are internal to a while_loop

### ISSUE THREAD

**Issue by rdipietro** [tensorflow/tensorflow#5412]
I'd like to manipulate the gradients that are internal to a while loop. The actual purpose is to view / manipulate recurrent-neural-network behavior, but here is a simplified example to get the idea across:

```
t_0 = tf.constant(0)
x_0 = tf.constant(1.0)
size = 3
x_ta = tf.TensorArray(tf.float32, size=size)

def cond(t, x_prev, x_ta):
    return tf.less(t, size)

def body(t, x_prev, x_ta):
    x = 2*x_prev
    x_ta = x_ta.write(t, x)
    return t + 1, x, x_ta

_, _, x_ta = tf.while_loop(cond, body, [t_0, x_0, x_ta])
x = x_ta.pack()

loss = x[size-1]

dloss_dx, = tf.gradients(loss, x)

with tf.Session() as sess:
    sess.run(tf.initialize_all_variables())
    print(sess.run(dloss_dx))
```

In this example, `x[t] = 2*x[t-1]` for all `t`, so we'd like `dloss_dx` to be `[4., 2., 1.]`. We instead get `[ 0.  0.  1.]`. I don't think this is a bug: we're actually taking the derivative with respect to the pack operation, which `x[2]` doesn't actually depend on.

Please correct me if I'm wrong, but I think it's currently impossible to obtain `[4., 2., 1.]` in the general case. (One hack here would be to compute gradients within the `while_loop` and then form a cumulative product corresponding to the chain rule, but this won't work with vector-valued `x[t]`'s because there's no way to get the Jacobians).

But the gradients with respect to each of these time steps must live somewhere. Can we add a property to the `TensorArray` class to be able to obtain these?

**Comment by yuanbyu** [comment]
I am not sure I understand this.  In your example, you seem to be doing tf.gradients(x[size-1], x) for some tensor x. So it should have nothing to do with the while loop or TensorArray?

**Comment by rdipietro** [comment]
Sorry.. let me try to use a simple example to show what I need:

```
x = [tf.constant(1.0)]
for _ in xrange(2):
    x.append(2*x[-1])

loss = x[-1]
dloss_dx = tf.gradients(loss, x)

with tf.Session() as sess:
    print(sess.run(dloss_dx))
```

This will print `[4., 2., 1.]`. This is what I need, but when using a `while_loop` instead of an explicit node-building loop. As far as I can tell, this is not currently possible, and I am requesting functionality to do this.

My thought is that these internal gradients must exist somewhere in memory when we call `sess.run`, because otherwise we couldn't be able to backpropagate through the loop, so I'm hoping it's as easy as adding functionality to read them out, sort of like a `TensorArray`'s `read` but for gradients instead of values.

Does that make sense?

**Comment by rdipietro** [comment]
Ping. Note that [tensorflow/python/ops/tensor_array_grad.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/tensor_array_grad.py) has a lot of sparsely-documented code that seems relevant, which was largely written by @ebrevdo. @ebrevdo would you be willing to share some insight on this?

To recap: Would like to add functionality to TensorFlow to read the per-time-step gradients that are internal to a `TensorArray`.

**Comment by rdipietro** [comment]
Ping. I'd really like to know if this is feasible, which isn't apparent from looking at the `TensorArray` source.

**Comment by rdipietro** [comment]
Friendly ping

**Comment by rdipietro** [comment]
Friendly ping @yuanbyu @ebrevdo @girving

**Comment by tfboyd** [comment]
Closing due to lack of activity.  Apologies that this did not come to a conclusion.  If the issue still exists or you still have a question please open a new issue here or on stackoverflow.
